lib/contextual_logger/logger_with_context.rb

```diff
@@ -36,6 +38,24 @@ def write_entry_to_log(severity, timestamp, progname, message, context:)
       @logger.write_entry_to_log(severity, timestamp, progname, message, context: merged_context)
     end
 
+    def normalize_context(context)
+      if warn_on_string_keys(context)
+        context.deep_symbolize_keys
+      else
+        context
+      end
+    end
+
+    def warn_on_string_keys(context)
+      if deep_key_has_string?(context)
+        ActiveSupport::Deprecation.warn('Context keys must use symbols not strings. This will be asserted as of contextual_logger v1.0.0')
+      end
+    end
+
+    def deep_key_has_string?(hash)
+      hash.keys.any? { |key| key.is_a? String } || hash.any? { |key, value| deep_key_has_string?(value) if value.is_a? Hash }
```

Comment:
@yinonrousso Also, our Ruby coding standards say to use parentheses around arguments unless it's a Domain Specific Language.

https://github.com/Invoca/style-guide/tree/master/ruby#omit-parentheses-around-parameters-for-methods-that-are-part-of-an-internal-dsl

Also, we only use trailing `if`/`unless` if it's a rare footnote that normally executes the preceding code.

https://github.com/Invoca/style-guide/tree/master/ruby#only-use-trailing-ifunless-when-they-are-rare-footnotes-that-can-be-ignored-in-the-usual-go-right-case

And the code is jammed up on one line. Here it is spread out a bit:
```
hash.keys.any? { |key| key.is_a?(String) } ||
   hash.any? { |key, value| (value.is_a?(Hash) && deep_key_has_string?(value) }
```
That's the same as:
```
hash.any? { |key, value| key.is_a?(String) } ||
   hash.any? { |key, value| (value.is_a?(Hash) && deep_key_has_string?(value) }
```
There's obviously no need to walk the hash twice!
```
hash.any? do |key, value|
  key.is_a?(String) } ||
    (value.is_a?(Hash) && deep_key_has_string?(value))
end
```

Reply:
Thank you, I'll make the appropriate changes 👍